spec/integration/blocked_hotlinked_image_spec.rb

```diff
@@ -0,0 +1,138 @@
+      expect(post.cooked).not_to have_tag("video source[src]")
+      expect(post.cooked).to have_tag("video source", with: { PrettyText::BLOCKED_HOTLINKED_SRCSET_ATTR => srcset })
+    end
+
+    it "blocks hotlinked audio" do
+      post = Fabricate(:post, raw: "![alt text|audio](#{hotlinked_url})")
+      expect(post.cooked).not_to have_tag("audio source[src]")
+      expect(post.cooked).to have_tag("audio source", with: { PrettyText::BLOCKED_HOTLINKED_SRC_ATTR => hotlinked_url })
+    end
+
+    it "blocks hotlinked onebox content when cached" do
+      post = Fabricate(:post, raw: "#{onebox_url}")
+      expect(post.cooked).not_to have_tag("img[src]")
+      expect(post.cooked).to have_tag("img", with: { PrettyText::BLOCKED_HOTLINKED_SRC_ATTR => hotlinked_url })
+    end
+
+    it "allows absolute URLs" do
+      src = "/assets/images/blah.png"
+      post = Fabricate(:post, raw: "![](#{src})")
+      expect(post.cooked).to have_tag("img", with: { src: src })
+    end
+
+    it "allows data URIs" do
+      src = "data:image/png;base64,abcde"
+      post = Fabricate(:post, raw: "![](#{src})")
+      expect(post.cooked).to have_tag("img", with: { src: src })
+    end
+
+    it "allows exceptions" do
+      post = Fabricate :post, raw: <<~RAW
+        ![](https://example.com)
+        ![](https://example.com/myimage.png)
+        ![](https://example.com.malicious.com/myimage.png)
+        ![](https://malicious.invalid/https://example.com)
+      RAW
+
+      post.rebake!
+      expect(post.cooked).not_to have_tag("img[src]")
+
+      SiteSetting.block_hotlinked_media_exceptions = "https://example.com"
+
+      post.rebake!
+      post.reload
+      expect(post.cooked).to have_tag("img", with: { "src" => "https://example.com" })
+      expect(post.cooked).to have_tag("img", with: { "src" => "https://example.com/myimage.png" })
+      expect(post.cooked).to have_tag("img", with: { PrettyText::BLOCKED_HOTLINKED_SRC_ATTR => "https://example.com.malicious.com/myimage.png" })
+      expect(post.cooked).to have_tag("img", with: { PrettyText::BLOCKED_HOTLINKED_SRC_ATTR => "https://malicious.invalid/https://example.com" })
+    end
+  end
+
+  context "with hotlinked media blocked, with post-processing" do
+    before do
+      SiteSetting.block_hotlinked_media = true
+      Jobs.run_immediately!
+      Oneboxer.stubs(:onebox).returns("<aside class='onebox'><img src='#{hotlinked_url}'></aside>")
+    end
+
+    it "renders placeholders for all media types" do
+      post = Fabricate :post, raw: <<~RAW
+        <img src='#{hotlinked_url}'>
+
+        ![alt text|video](#{hotlinked_url})
+
+        ![alt text|audio](#{hotlinked_url})
+
+        #{onebox_url}
+      RAW
+      post.rebake!
+      post.reload
+      expect(post.cooked).not_to have_tag("img")
+      expect(post.cooked).not_to have_tag("video")
+      expect(post.cooked).not_to have_tag("audio")
+      expect(post.cooked).to have_tag(".blocked-hotlinked-placeholder", count: 4)
```

Comment:
I would personally test in greater detail here. Instead of just asserting the tag exists, I would test for how `post.cooked` would look like after processing since a fair amount of code has been added to the processing of a post.

Reply:
Testing the whole content of post.cooked would look something like

```ruby
expect(post.cooked.to eq(<<~HTML)
  <p><a data-blocked-hotlinked-src="http://example.com/images/2/2e/Longcat1.png" width="" height="" href="http://example.com/images/2/2e/Longcat1.png" class="blocked-hotlinked-placeholder" title="Media hosted on another site. Click to open in a new tab." rel="noopener nofollow ugc"><svg class="fa d-icon d-icon-link svg-icon" aria-hidden="true"><use href="#link"></use></svg><span class="notice">External Media</span></a></p>
  <p></p><a href="http://example.com/images/2/2e/Longcat1.png" class="blocked-hotlinked-placeholder" title="Media hosted on another site. Click to open in a new tab." rel="noopener nofollow ugc"><svg class="fa d-icon d-icon-link svg-icon" aria-hidden="true"><use href="#link"></use></svg><span class="notice">External Media</span></a><p></p>
  <p><a href="http://example.com/images/2/2e/Longcat1.png" class="blocked-hotlinked-placeholder" title="Media hosted on another site. Click to open in a new tab." rel="noopener nofollow ugc"><svg class="fa d-icon d-icon-link svg-icon" aria-hidden="true"><use href="#link"></use></svg><span class="notice">External Media</span></a></p>
  <aside class="onebox"><a data-blocked-hotlinked-src="http://example.com/images/2/2e/Longcat1.png" width="" height="" href="http://example.com/images/2/2e/Longcat1.png" class="blocked-hotlinked-placeholder" title="Media hosted on another site. Click to open in a new tab." rel="noopener nofollow ugc"><svg class="fa d-icon d-icon-link svg-icon" aria-hidden="true"><use href="#link"></use></svg><span class="notice">External Media</span></a></aside>
HTML
```
IMO that's not easy to read, and not really very helpful. Any tiny change to the core logic will require re-copy-pasting the HTML.

That said, I will add a few more checks to make sure each placeholder has an `a[href]` to the correct place.

Reply:
Updated in 6ca0eb136e70a7a4484af2a732aa4f9aed160fa9

Reply:
Looks good :+1: